extensions/common/gcp/gcp-core/src/main/java/org/eclipse/dataspaceconnector/gcp/core/iam/IamServiceImpl.java

```diff
@@ -0,0 +1,198 @@
+import com.google.cloud.iam.credentials.v1.IamCredentialsClient;
+import com.google.cloud.iam.credentials.v1.ServiceAccountName;
+import com.google.common.collect.ImmutableList;
+import com.google.iam.admin.v1.CreateServiceAccountRequest;
+import com.google.iam.admin.v1.ProjectName;
+import com.google.iam.admin.v1.ServiceAccount;
+import com.google.protobuf.Duration;
+import org.eclipse.dataspaceconnector.gcp.core.common.GcpException;
+import org.eclipse.dataspaceconnector.gcp.core.common.ServiceAccountWrapper;
+import org.eclipse.dataspaceconnector.gcp.core.storage.GcsAccessToken;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class IamServiceImpl implements IamService {
+
+    public static final ImmutableList<String> OAUTH_SCOPE = ImmutableList.of("https://www.googleapis.com/auth/cloud-platform");
+    private static final long ONE_HOUR_IN_S = TimeUnit.HOURS.toSeconds(1);
+    private final String gcpProjectId;
+    private final Supplier<IAMClient> iamClientSupplier;
+    private final Supplier<IamCredentialsClient> iamCredentialsClientSupplier;
+    private final Monitor monitor;
+
+    private IamServiceImpl(Monitor monitor,
+                           String gcpProjectId,
+                           Supplier<IAMClient> iamClientSupplier,
+                           Supplier<IamCredentialsClient> iamCredentialsClientSupplier
+    ) {
+        this.monitor = monitor;
+        this.gcpProjectId = gcpProjectId;
+        this.iamClientSupplier = iamClientSupplier;
+        this.iamCredentialsClientSupplier = iamCredentialsClientSupplier;
+    }
+
+    @Override
+    public ServiceAccountWrapper getOrCreateServiceAccount(String serviceAccountName, String serviceAccountDescription) {
+        var requestedServiceAccount = ServiceAccount.newBuilder()
+                .setDisplayName(serviceAccountName)
+                .setDescription(serviceAccountDescription)
+                .build();
+        var request = CreateServiceAccountRequest.newBuilder()
+                .setName(ProjectName.of(gcpProjectId).toString())
+                .setAccountId(serviceAccountName)
+                .setServiceAccount(requestedServiceAccount)
+                .build();
+
+        try (var client = iamClientSupplier.get()) {
+            var serviceAccount = client.createServiceAccount(request);
+            monitor.info("Created service account: " + serviceAccount.getEmail());
+            return new ServiceAccountWrapper(serviceAccount.getEmail(), serviceAccount.getName(), serviceAccountDescription);
+        } catch (ApiException e) {
+            if (e.getStatusCode().getCode() == StatusCode.Code.ALREADY_EXISTS) {
+                return getServiceAccount(serviceAccountName, serviceAccountDescription);
+            }
+            monitor.severe("Unable to create service account: \n" + e);
```

Comment:
the exception should be passed as the second parameter so the whole stacktrace will be printed (useful for severe logs)